### PR TITLE
Correct scanning for literals such as 1.0e-999999999999999999999 (should convert to zero)

### LINKED ISSUE #6109: Error in scanning "1.0e-99999999999999999999999"; should convert to zero
The literal `1.0e-99999999999999999999999` should scan as the value `0.0`, but it acts as if it is an overflow (compile-time error).

### PR DESCRIPTION
**Pull request by leppie**
Fixes https://github.com/dotnet/roslyn/issues/6109.

Please advise if tests are needed and where they meant to be.

/cc @gafter

### CODE REVIEW

File: src/Compilers/Core/Portable/RealParser.cs [hunk 1]
```diff
@@ -358,14 +358,24 @@ public static DecimalFloatingPointString FromSource(string source)
...
+                                exponent -= exponentMagnitude;
+                            }
+                            else
+                            {
+                                exponent += exponentMagnitude;
+                            }
+                        }
                     }
-                    else
+                    catch (OverflowException)
```
Comment (gafter) [inline src/Compilers/Core/Portable/RealParser.cs]: Please do not use exception handling.

**Comment by gafter** [comment]
I recommend adding tests in https://github.com/dotnet/roslyn/blob/master/src/Compilers/Core/CodeAnalysisTest/RealParserTests.cs

**Comment by leppie** [comment]
This is turning into an endless battle... :disappointed: 

Reporting `CS0594` for invalid floating point literals seems incorrect. These should be rejected by the tokenizer/lexer already.

**Comment by leppie** [comment]
Too much spaghetti logic involved to fix this now, will maybe try again at a later stage.

**Comment by gafter** [comment]
@leppie this is part of the scanner.

**Comment by leppie** [comment]
@gafter The scanner is allowing nonsense to pass (https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Portable/Parser/Lexer.cs#L1032) and there is no error that I can see that allows for indicating an invalid real literal except CS0594 which is also nonsense for these cases. If there was at least some error like Invalid Number like for integers, that could be applied, but this does not seem the case. I also dont think I am in a position to create a new error message for this case. Overall, rejection of invalid syntax in C# seems very poor with most errors being too specific.

**Pull request closed without merging**